### ISSUE TITLE
If an applicant enters a number value larger than the max long value, they get a 400

### ISSUE THREAD

**Issue by natsid** [seattle-uat/civiform#815]
We should instead ensure the number is no bigger than the max long. And if it is, return a helpful error message.

**Comment by natsid** [comment]
This bug blocks story acceptance of #377.

**Comment by shanemc-goog** [comment]
A validation.ts method should be straightforward, or configuring the input for existing HTML attributes.

**Comment by th174** [comment]
Is this resolved by https://github.com/civiform/civiform/pull/1706

**Comment by bion** [comment]
@th174 no, #1706 only addressed the client-side experience. For this to be done the server also needs to handle large values appropriately.

**Comment by BryanGarciaCodes** [comment]
I would like to work on this issue. Is this available??

**Comment by melanie-exygy** [comment]
Hi @BryanGarciaCodes, thanks for your interest in contributing to CiviForm!  Here is some info on how to start contributing:  [Technical contribution guide](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#outside-contributors-not-part-of-exygy-google-or-a-civic-entity).  Please introduce yourself on our Slack channel [#community-contributors](https://civiform.slack.com/archives/C07S691HDU0).

**Comment by SamRaymer** [comment]
It looks like the behavior gets weird at fewer digits than this when trying a form with a currency field, curious about how to determine the actual max and min acceptable values.

**Comment by melanie-exygy** [comment]
> It looks like the behavior gets weird at fewer digits than this when trying a form with a currency field, curious about how to determine the actual max and min acceptable values.

I think it's very unlikely governments would ever need such a high max/min.  @shreyachatterjee00 , would you be able to help @SamRaymer determine some more realistic default max and min for currency questions?

**Comment by shreyachatterjee00** [comment]
Can we do something like `one billion` in either direction? @melanie-exygy

**Comment by melanie-exygy** [comment]
> Can we do something like `one billion` in either direction? [@melanie-exygy](https://github.com/melanie-exygy)

@SamRaymer would you mind giving 10 digits a try to see if you get that same weird behavior?

**Comment by SamRaymer** [comment]
@melanie-exygy It works fine with 10 digits. I'll start a change to cap at one billion. Thanks all!

**Comment by rockycodes** [comment]
@shreyachatterjee00 @SamRaymer (fyi @swatkat1 ) we need to revisit the logic here because the one billion cap won't work for how seattle uses number questions (they use them for ids sometimes). rather than limiting to 10 digits, we should limit closer to the number of digits a min/max long can be.

**Comment by msprenke** [comment]
This is an issue for us. We use the number field for trusted intermediaries to enter the card number for Freshbucks when they hand out cards.  Example card number 63681171008117130 which is 63,681,171,008,117,130. There is a likely hood that a run of cards could be up to 999,999,999,999,999,999

**Comment by rockycodes** [comment]
@SamRaymer can we make this fix just for the number question and use the max/min long as the limits?
Max long: 9,223,372,036,854,775,807
Min long: -9,223,372,036,854,775,808

**Comment by SamRaymer** [comment]
> [@SamRaymer](https://github.com/SamRaymer) can we make this fix just for the number question and use the max/min long as the limits? Max long: 9,223,372,036,854,775,807 Min long: -9,223,372,036,854,775,808

@rockycodes It looks like there are already errors for those values. The JSON parser gives an error for values that don't fit into a long but right now the copy is slightly off
> Error: Number must be a positive whole number and can only contain numeric characters 0-9.

Should I add some kind of regex in the parser to catch a number that is just too big?

**Comment by rockycodes** [comment]
@shreyachatterjee00 @myfriendgu thoughts on copy changes? looks like we already catch numbers that are too large. 

for currency questions, can one of you figure out what is the current max/min and if we want to change it? seems like number questions and currency questions should not have the same max/min.

**Comment by myfriendgu** [comment]
I'm trying to understand what we want to limit so we can figure out an appropriate error message for users.

For number questions, the error message above says "Error: Number must be a positive whole number and can only contain numeric characters 0-9."

- Does it need to be positive whole numbers only? Are we ok with ruling out decimal points? Seems like there might be a use case for this
- Can users enter negatives? If not, why are we defining the min long as -9,223,372,036,854,775,808? Shouldn't the min long be 0?
- Is there a reason the max and min length is so specific? Why can't it be 999,999,999,999,999,999?


For currency questions
- I assume we **are** supporting factional numbers, so we can take either decimal points or commas... is that correct?
- Is there a reason why we need a different error for currency questions? The only reason I can think of to distinguish between the two is if the regular number questions can only take whole numbers and also negative numbers (although I'm guessing there could be a use case for currency questions to also take negatives).

cc @SamRaymer and @rockycodes

**Comment by myfriendgu** [comment]
This is the current error for currency: 

<img width="461" height="134" alt="Image" src="https://github.com/user-attachments/assets/d0a73db3-8da1-4a8e-9027-ae9ddae0c491" />

Obviously, the copy needs to be updated to make more sense, which I'll work on

**Comment by SamRaymer** [comment]
@myfriendgu 
 
>  Does it need to be positive whole numbers only? Are we ok with ruling out decimal points? Seems like there might be a use case for this

The database is flexible, would be possible to make the change. @rockycodes opinions on this one?
 
>  Can users enter negatives? If not, why are we defining the min long as -9,223,372,036,854,775,808? Shouldn't the min long be 0?

Negatives are not allowed in the current implementation, but we could change that. Leaving this to @rockycodes as well.

>   Is there a reason the max and min length is so specific? Why can't it be 999,999,999,999,999,999?

This is based on the maximum value of a 64 bit number. We could change it to `BigInteger` or something if we need to go even bigger.


> For currency questions
> I assume we **are** supporting factional numbers, so we can take either decimal points or commas... is that correct?

We do send fractional dollars (technically we send the total number of cents) but they aren't formatted with the comma separators.

> Is there a reason why we need a different error for currency questions? The only reason I can think of to distinguish between the two is if the regular number questions can only take whole numbers and also negative numbers (although I'm guessing there could be a use case for currency questions to also take negatives).

Right now the distinction mostly has to do with formatting the field to support two decimal places.

**Comment by SamRaymer** [comment]
> for currency questions, can one of you figure out what is the current max/min and if we want to change it? seems like number questions and currency questions should not have the same max/min.

Right now the max for numbers is the max value of a `long`, and the max for currency is the max value of a `long` in cents.

**Comment by SamRaymer** [comment]
@rockycodes ^

**Comment by rockycodes** [comment]
@SamRaymer Jia and Shreya talked it over and will have copy changes for you later today. We're going to avoid changing the scope to any changes to how the field works.

I tested out currency and I see what you mean about it behaving strangely. Will you make a separate issue for that? We can just address the copy updates in this bug.

**Comment by myfriendgu** [comment]
@SamRaymer Here's the updated text for each type of error alert:

**Regular number** 
- _Original_
Error: Number must be a positive whole number and can only contain numeric characters 0-9.
- _Updated_
Error: Number must be a positive whole number and can only contain numeric characters 0-9. Maximum length is 19 characters long.

**Currency number**
- _Original_
Error: Currency must be one of the following formats: 1000 1,000 1000.30 1,000.30
- _Updated_
Error: Currency must have two numeric characters (0-9) after the decimal. Maximum length is 19 characters long.
